okio/src/nativeMain/kotlin/okio/DataProcessor.kt

```diff
@@ -55,10 +55,188 @@ internal abstract class DataProcessor : Closeable {
   var closed: Boolean = false
     protected set
 
+  /** True if the content is self-terminating and has reached the end of the stream. */
+  var finished: Boolean = false
+    internal set
+
   /**
    * Returns true if no further calls to [process] are required to complete the operation.
-   * Otherwise, make space available in [target] and call [process] again.
+   * Otherwise, make space available in [target] and call this again.
    */
   @Throws(ProtocolException::class)
   abstract fun process(): Boolean
+
+  /** True if calling [process] may produce more output without more input. */
+  private var callProcess = false
+
+  /**
+   * Consume [sourceExactByteCount] bytes from [source], writing any amount of output to [target].
+   *
+   * Note that 0 is a valid number of bytes to process, and this will cause flush and finish blocks
+   * to be written. For such 0-byte writes [source] may be null.
+   */
+  @Throws(IOException::class)
+  fun writeBytesFromSource(
+    source: Buffer?,
+    sourceExactByteCount: Long,
+    target: BufferedSink,
+  ) {
+    check(!closed) { "closed" }
+
+    var byteCount = 0
+    while (true) {
+      val sourceHead = prepareSource(source, sourceExactByteCount - byteCount)
+      val targetTail = prepareTarget(target)
+      try {
+        callProcess = !process()
+      } finally {
+        byteCount += updateSource(source, sourceHead)
+        updateTarget(target, targetTail)
+      }
+
+      // If we've produced a full segment, emit it. This blocks writing to the target.
+      target.emitCompleteSegments()
```

Comment:
Better to do this inside the loop or outside? I would have thought outside since this is called once per `write` on the source.

Reply:
Good question! I think inside is correct specifically for inflate + deflate.

Delivering data one-segment-at-a-time is pretty balanced:
 - limit how much memory it holds. If you have a bunch of these on a web server, you don‘t have to worry about giant allocations when data decompresses tremendously!
 - reduce latency. We can deliver data to the user (or network, or whatever) as soon as it’s ready
 - still get some benefits from batch processing. We do expensive operations like syscalls once every 8 KiB to amortize the cost

We’d get maximum efficiency (fewer syscalls) if we delivered more segments at a time. But I think the real optimization there is to do a once-per-decade review of Segment.SIZE and see if 8 KiB makes as much sense in 2024 as it did in 2014.

Reply:
(Last time I considered dynamic segment sizes I concluded it was a bad fit. Pooling gets weirder, for one.)

Reply:
Yeah I guess I was more thinking as to whether it made sense for a single `Sink.write` call to result in 500 `Sink.write`s beneath this, or whether it should be a single call with a multi-segment buffer. I suppose in practice the compression level isn't going to be astronomical so it doesn't _really_ matter.